[#12329] Refactoring of sortable tables - Sessions table

Zxun2 opened the pull request:
Part of #12329 
  
## Outline of Solution

**Note:** This refactoring works are only on the `sessions-table` component which affects `Instructor Home` and `Instructor Sessions`. It should be easy enough to integrate `sessions-table` later on to other sessions-related tables (e.g. `recycle-bin-table`).

Previously, the data on the table component relies on the information provided by `x-tableRowModels` -- meaning that angular will "capture" and "trigger" a UI change in the DOM should there be any changes to `x-tableRowModels`. The usage of the `sortable-table` component meant that `rowData` is now responsible for "triggering" this UI change, which results in an indirect dependency to `x-tableRowModels` because there is a need to sync `x-tableRowModels` when a change in `rowData` is detected, and vice versa. This dependency is likely to be justified, as this is a scalable way to ensure that the `sortable-table` component is as versatile as possible. However, more caution is needed when developing as the syncing of two states is very bug prone. E2E testing is very beneficial here.

### Sortable Tables

The existing component is very well defined. However, there is currently no way to customize the stylings of the table, so I added that capability. Generally, there are two supported stylings (as i have observed): 

| style 1 | style 2|
| --- | --- |
| ![image](https://github.com/TEAMMATES/teammates/assets/63457492/202b6605-faf1-4517-9f2f-36cb2dd1d3ea)| ![image](https://github.com/TEAMMATES/teammates/assets/63457492/356bbc46-ce93-4f76-9dbf-d49b567d5c6c) |

I have also modified `ColumnData` to add two new attributes, `alignment` and `headerClass`. 
- `alignment` denotes the text-position in the cell and supports the conventional, `'start' | 'center' | 'end'`. 
- `headerClass` is useful for any additional stylings.

```ts
/**
 * Column data for sortable table
 */
export interface ColumnData {
  :
  alignment?: 'start' | 'center' | 'end'; // defaults to start via getAlignment method
  headerClass?: string; // additional styles
}
```
and the `componentData` of `SortableTableCellData` is now a function that accepts an index, indicating it's row's position in the `sortable-table`.

```ts
componentData: (idx: number) => Record<string, any>, // @Input values for component
```

Additionally, the component now emits a `sortEvent` (whenever `sortRows` is called), which is required to synchronize the state of `rowsData` and `x-tableRowModels`.

### Sessions Table 

The entire sessions table is now completely replaced with the `sortable-table` component. To support the current stylings of `sessions-table`, I created the following components: 
- `cell-with-tooltip`
- `cell-with-group-buttons`
- `cell-with-response-rate` 

These are the two functions responsible for creating the rows and columns:

```ts
  /**
   * Creates the column data for the table.
   * 
   * @param config Contains the information to create a column data.
   * @returns An array containing a column's data for the table
   *  if the column is to be shown, otherwise an empty array.
   */
  createColumnData(config: SessionsTableColumnData): ColumnData[] {
    if (!(config?.columnType === undefined) && !this.columnsToShow.includes(config.columnType!)) {
      return [];
    }

    const columnData: ColumnData = {
      header: config.header,
      ...(config.sortBy && { sortBy: config.sortBy }),
      ...(config.headerToolTip && { headerToolTip: config.headerToolTip }),
      ...(config.alignment && { alignment: config.alignment }),
      ...(config.headerClass && { headerClass: config.headerClass }),
    };

    return [columnData];
  }

  /**
   * Creates the row data for the table.
   * 
   * @param config Contains the information to create a row data.
   * @returns An array containing a row's data for the table if the row is to be shown,
   * otherwise an empty array.
   */
  createRowData(config: SessionsTableRowData): SortableTableCellData[] {
    if (!(config?.columnType === undefined) && !this.columnsToShow.includes(config.columnType!)) {
      return [];
    }

    const rowData: SortableTableCellData = {
      ...(config.value && { value: config.value }),
      ...(config.displayValue && { displayValue: config.displayValue }),
      ...(config.customComponent && { customComponent: config.customComponent }),
      ...(config.style && { style: config.style }),
    };

    return [rowData];
  }

```

## Screenshots

It looks exactly like it currently does...

Zxun2 commented on src/e2e/java/teammates/e2e/pageobjects/Browser.java:
```diff
@@ -209,6 +209,7 @@ private WebDriver createWebDriver() {
             edgePrefs.put("download.prompt_for_download", false);
             EdgeOptions options = new EdgeOptions();
             options.setExperimentalOption("prefs", edgePrefs);
+            options.addArguments("--remote-allow-origins=*");
```
Part of #12035.

Fix EdgeBrowser setup for E2E Tests.

Zxun2 commented on src/e2e/java/teammates/e2e/cases/InstructorFeedbackSessionsPageE2ETest.java:
```diff
@@ -88,7 +88,6 @@ public void testAll() {
 
         ______TS("verify loaded data");
         FeedbackSessionAttributes[] loadedSessions = { openSession, closedSession };
-        feedbackSessionsPage.sortByCourseId();
```
No longer needed to sort by course id because table is already sorted in ASC

Zxun2 commented on src/e2e/java/teammates/e2e/pageobjects/InstructorHomePage.java:
```diff
@@ -151,7 +151,7 @@ private String getCourseDetails(int courseTabIndex) {
 
     private WebElement getSessionsTable(int courseTabIndex) {
-        return getCourseTab(courseTabIndex).findElement(By.className("sessions-table"));
+        return getCourseTab(courseTabIndex).findElement(By.id("sessions-table-" + courseTabIndex));
```
There may be multiple tables in the `InstructorHomePage` (due to having multiple courses) so we should try to identify each table using a unique id.

Zxun2 commented on src/web/app/components/sortable-table/sortable-table.component.ts:
```diff
@@ -53,25 +60,28 @@ export class SortableTableComponent implements OnInit, OnChanges {
 
   ngOnInit(): void {
-    this.tableRows = this.rows.slice(); // Shallow clone to avoid reordering original array
+    this.tableRows = this.rows;
```
Not sure if this is a reasonable thing to do, but i need to mutate the original array to ensure that the row state are synced across `sortable-table`, `sessions-table` and `instructor-home-page`/`instructor-sessions-page`.

weiquu replied:
Thanks - checked and didn't find any issues as well

Zxun2 commented on src/web/app/components/sortable-table/sortable-table.component.ts:
```diff
@@ -87,28 +97,32 @@ export class SortableTableComponent implements OnInit, OnChanges {
       return;
     }
-
+    this.sortEvent.emit({ sortBy, sortOrder: this.sortOrder });
```
`sortable-table` emits both the column to sort by and the sorting order. This is necessary to sync the `tableRowsModels` and `rowsData` together.

Zxun2 commented on src/web/app/pages-instructor/instructor-sessions-page/instructor-sessions-page.component.ts:
```diff
@@ -124,8 +135,6 @@ export class InstructorSessionsPageComponent extends InstructorSessionModalPageC
   isSessionEditFormExpanded: boolean = false;
 
   sessionsTableRowModels: SessionsTableRowModel[] = [];
-  sessionsTableRowModelsSortBy: SortBy = SortBy.NONE;
-  sessionsTableRowModelsSortOrder: SortOrder = SortOrder.ASC;
```
No longer necessary to track the sorting state, since the `sortable-table` component now emits the sorting state. 
See [comment](https://github.com/TEAMMATES/teammates/pull/12501/files#r1254558184).

Zxun2 commented:
Apologies for taking rather long on this. Had some trouble dealing with the failing e2e testcases🥲

Zxun2 commented:
@weiquu Yup, that's right!

weiquu commented:
Thanks @Zxun2! Do take note of the following:
- Try not to combine fixes for multiple issues in 1 PR, unless they are tightly related (see [here](https://teammates.github.io/teammates/process.html#step-2-start-clean-from-a-new-branch)). You can make a separate PR for the E2E EdgeBrowser setup
- As mentioned further down in the same page (point 3 [here](https://teammates.github.io/teammates/process.html#step-4-submit-a-pr)), do ensure that "no unrelated changes are introduced in the branch. This includes unnecessary formatting changes." These are mainly in the `.html` files, but there are a few instances in the `.ts` files as well. These changes make it difficult to review an already large PR, since we have to determine which are merely formatting changes and which are not. Not sure if it's possible to reverse these changes, given that the code readability is also not necessarily increased (e.g. an element now takes up more lines than it usually does)
- Do make the title a bit more descriptive, e.g. "Refactoring of sortable tables - sessions table". I'll also update the issue description to use a checklist

(Have requested for a couple more reviewers than usual given the size of the PR and how it will probably influence the rest of the refactorings)

weiquu commented:
Hi @Zxun2, can I check the rationale for changing the default sort behaviour for the Instructor Home Page to session name ascending? I would have thought sorting by session end date descending would be more useful (e.g. for courses with many sessions, I want to see sessions that are ongoing / coming up over sessions where the end date has passed)

Also noted for the instructor sessions page, the default behaviour was changed from descending to ascending - think this is fine!

Overall, everything looks great (: Currently working through the code itself - will update in a bit

Zxun2 commented:
Hey @weiquu, thanks for the review! I can't recall the exact rationale, but it was along the lines of, if it was 
- `ASC`, then sessions that are closed will be placed on top (because they have the earliest end date)
- `DSC`, then sessions with the latest ending date will be placed on top (i thought this didn't make much sense since i expect that sessions that are open and ending soon should be on top instead.)

so i settled on sorting by `Session Name` in the end 😅 I have reverted that change to the original (which is `Session End Date DSC`, and if necessary, we can implement the above in another PR. 

>  I also noticed some changes to contribution-question-statistics.component.ts - are those changes relevant to the PR? Was a bit surprised to notice the file there

I changed the `SortableTableCellData` interface which necessitates changes to `contribution-question-statistics.component.ts` which is already using the `sortable-table` component. The `componentData` attribute was originally `Record<string, any>` but it is now a function which accepts an `idx` which denotes the row it belongs to.

weiquu commented on src/web/app/components/sortable-table/sortable-table.component.html:
```diff
@@ -1,26 +1,29 @@
-<table class="table table-bordered table-striped margin-bottom-20px">
```
Seems that removing this line has caused some complications with the other components using the `sortable-table`:
<img width="1200" alt="Screenshot 2023-07-23 at 12 48 12 AM" src="https://github.com/TEAMMATES/teammates/assets/48304907/54cbd748-1e14-45e5-ba31-5e37ac841f5e">

Think it affects most of the question statistics components... so either
- Add the `margin-bottom-20px` back in, and handle it separately in the instructor home and sessions pages
- Make it another customisable part of the `sortable-table` component...
- Add the margins back in each of the other question statistics components (and remove the now redundant styling from the `sortable-table.component.scss` file)

Not too sure which is the best approach, though I think the last one is easiest and it also makes sense that the margin after the table is left to each component to handle.

Zxun2 replied:
I agree with the last approach as well! I will make the changes shortly 👍🏻

weiquu commented:
Thanks @Zxun2 for the change and the info! Just one small issue to fix (:

(Taking this opportunity to ping @jasonqiu212 @domlimm @cedricongjh @EuniceSim142 - if a couple of you could review this PR soon that would be great)

weiquu reviewed (approved):
LGTM! Thanks for the changes and the great work (:

jasonqiu212 commented on src/web/app/components/question-types/question-statistics/rubric-question-statistics.component.html:
```diff
@@ -45,7 +45,7 @@
     </div>
     <div class="row">
-      <div id="rubric-recipient-statistics-overall" class="col-sm-12 table-responsive">
+      <div id="rubric-recipient-statistics-overall" class="col-sm-12 table-responsive mb-5">
```
May I ask what's the rationale for using `mb-5` here?

Zxun2 replied:
I forgot to change back after playing around with it 😅

jasonqiu212 commented on src/web/app/components/sessions-table/sessions-table.component.html:
```diff
@@ -1,140 +1,12 @@
+    [sortOrder]="sortOrder"
+    [headerColorScheme]="headerColorScheme"
+    (sortEvent)="sortSessionsTableRowModelsEventHandler($event)"
+    ></tm-sortable-table>
```
that's pretty satisfying haha

jasonqiu212 commented on src/web/app/components/sortable-table/sortable-table.component.ts:
```diff
@@ -44,6 +53,12 @@ export class SortableTableComponent implements OnInit, OnChanges {
+  tableId: string = '';
+
+  @Input()
+  headerColorScheme: SessionsTableHeaderColorScheme = SessionsTableHeaderColorScheme.BLUE;
```
Should `SessionsTableHeaderColorScheme` be renamed and moved elsewhere? Feels weird to use a color scheme defined for Sessions Table for all `sortable-table`s

Zxun2 replied:
Yup, sure! I have shifted it to the `sortable-table` component and updated the naming from `SessionsTableHeaderColorScheme` to `SortableTableHeaderColorScheme`.

jasonqiu212 reviewed (commented):
Outstanding work @Zxun2! Thanks for the huge refactor :) Just 2 small questions

jasonqiu212 reviewed (approved):
LGTM, thanks @Zxun2!
